llama/llama.go

```diff
@@ -692,31 +674,42 @@ func SchemaToGrammar(schema []byte) []byte {
 	return buf[:n]
 }
 
-type Sampler struct {
-	c *C.struct_llama_sampler
+type TokenData struct {
+	Id    int32
+	Logit float32
 }
 
-func NewGrammarSampler(vocab *Vocab, grammar string) *Sampler {
-	cGrammar := C.CString(grammar)
-	cRoot := C.CString("root")
-	defer C.free(unsafe.Pointer(cGrammar))
-	defer C.free(unsafe.Pointer(cRoot))
+type Grammar struct {
+	c *C.struct_llama_grammar
+}
 
-	sampler := &Sampler{c: C.llama_sampler_init_grammar(vocab.c, cGrammar, cRoot)}
+func (g *Grammar) AddTokenPiece(token uint32, piece string) {
```

Comment:
This should happen when we're instantiating an `ollama_vocab`, but that doesn't seem to belong as a member function of `Grammar`

Reply:
good callout - should be able to coalesce some of this and keep the go <> c interface simpler